spynnaker/pyNN/external_devices_models/spif_input_device.py

```diff
@@ -0,0 +1,296 @@
+        self.__m_vertex_mask = ((self.__key_mask << self.__neuron_bits) +
+                                (sub_mask << self.__index_shift))
+
+        # A dictionary to get vertex index from FPGA and slice
+        self.__index_by_slice = dict()
+
+        self.__pipe = pipe
+        self.__base_key = base_key
+        if self.__base_key is None:
+            self.__base_key = SPIFInputDevice.__n_devices
+        SPIFInputDevice.__n_devices += 1
+
+        # Generate the shifts and masks to convert the SPIF Ethernet inputs to
+        # output format
+        self.__input_mask = (1 << self.__neuron_bits) - 1
+        self.__input_shift = 0
+
+    def __is_power_of_2(self, v):
+        """ Determine if a value is a power of 2
+
+        :param int v: The value to test
+        :rtype: bool
+        """
+        return (v & (v - 1) == 0) and (v != 0)
+
+    def __incoming_fpgas(self, board_address, chip_coords):
+        """ Get the incoming FPGA connections
+
+        :rtype: list(FPGAConnection)
+        """
+        # We use every other odd link
+        return [FPGAConnection(SPIF_FPGA_ID, i, board_address, chip_coords)
+                for i in SPIF_INPUT_FPGA_LINKS]
+
+    def __outgoing_fpga(self, board_address, chip_coords):
+        """ Get the outgoing FPGA connection (for commands)
+
+        :rtype: FGPA_Connection
+        """
+        return FPGAConnection(
+            SPIF_FPGA_ID, SPIF_OUTPUT_FPGA_LINK, board_address, chip_coords)
+
+    def __fpga_index(self, fpga_link_id):
+        # We use every other odd link, so we can work out the "index" of the
+        # link in the list as follows
+        return (fpga_link_id - 1) // 2
+
+    @overrides(ApplicationFPGAVertex.get_incoming_slice_for_link)
+    def get_incoming_slice_for_link(self, link, index):
+        vertex_slice = super().get_incoming_slice_for_link(link, index)
+        self.__index_by_slice[link.fpga_link_id, vertex_slice] = index
+        return vertex_slice
+
+    @overrides(ApplicationFPGAVertex.get_machine_fixed_key_and_mask)
+    def get_machine_fixed_key_and_mask(self, machine_vertex, partition_id):
+        fpga_link_id = machine_vertex.fpga_link_id
+        vertex_slice = machine_vertex.vertex_slice
+        index = self.__index_by_slice[fpga_link_id, vertex_slice]
+        key = ((self.__base_key << self.__neuron_bits) |
+               index << self.__index_shift)
+
+        fpga = self.__fpga_index(fpga_link_id)
+
+        # Build the key from the components
+        fpga_key = key + fpga
+        fpga_mask = self.__m_vertex_mask | self.__spif_mask
+        return BaseKeyAndMask(fpga_key, fpga_mask)
+
+    @overrides(ApplicationFPGAVertex.get_fixed_key_and_mask)
+    def get_fixed_key_and_mask(self, partition_id):
+        n_key_bits = BITS_IN_KEY - self.__neuron_bits
+        key_mask = ((1 << n_key_bits) - 1) << self.__neuron_bits
+        return BaseKeyAndMask(self.__base_key << self.__neuron_bits, key_mask)
```

Comment:
Why does this method recalculate the mask rather than use the one from the init.

Reply:
Fixed